### PR TITLE
Unattended deployment on a vanilla image

### LINKED ISSUE #220: Deploy image in COS_ACTIVE in YIP
This card is about making a YIP plugin to deploy an image to COS_ACTIVE partition.

The image/channel to deploy from should be configurable. And also setting the COS_PASSIVE with same image or not should be configurable. This way it could be used for the first time deployment but also for upgrades.

Note the code from card could be reused for #108 or viceversa

### PR DESCRIPTION
**Pull request by davidcassany**
This PR allows unattended deployments on top of vanilla images by
using simple cloud init configuration files that can be injected to the
system as part of the user data of public providers.
    
This PR includes an `images/user-data.yaml` file as an example.
    
Fixes #220

### CODE REVIEW

**Comment by davidcassany** [comment]
This PR is rebased on top of #305, so it is blocked until the base gets merged.

**Review by mudler — APPROVED** [review]
Looks good here!

File: packages/installer/deploy.sh
Comment (Itxaka) [inline packages/installer/deploy.sh:10]: everything in here seems very similar to cos-installer. Could we extract most of this functions into a cos-functions file and then source it in both scripts?

Comment (mudler) [reply packages/installer/deploy.sh:10]: that sounds good, but I would rather do in a separate follow-up PR. The same applies for `cos-upgrade`, `cos-installer` and `cos-reset`

**Comment by mudler** [comment]
base was merged, removing blocked label

**Comment by mudler** [comment]
Can we add a small test here? Just to check we don't break cos-deploy accidentally

File: packages/cos/build.yaml
Comment (mudler) [inline packages/cos/build.yaml:24]: Any specific reason for re-adding it? In case it's needed, we also need to update the recovery bootargs

File: packages/installer/deploy.sh
Comment (davidcassany) [reply packages/installer/deploy.sh:10]: Yes I agree I would do it in a separate PR or even leave it as is until we start working on #108 IMHO all this scripts should be substituted by a client in the scope of #108. Probably a fast solution is to build a shell client so we can reuse many portions of the code and we not have to start from scratch, I am not very passionate about this option, but probably this is a good way to start splitting all these logic into pieces and compose some client skeleton.

File: packages/cos/build.yaml
Comment (davidcassany) [reply packages/cos/build.yaml:24]: Yes, because then at deploy time the labels can't be applied, it requires SE Linux binaries to apply the labels.

File: packages/installer/deploy.sh
Comment (davidcassany) [reply packages/installer/deploy.sh:10]: Indeed I am not very happy of this `cos-deploy` script, but I though that abusing of `cos-upgrade` to also cover this use case was making the behavior far from obvious (very small code changes were needed, but then the _upgrade_ behavior looked pretty strange to my eyes). That's why I opted for creating its own script by just copying the `cos-upgrade` code, removing many parts and applying few logic changes. I duplicated quite some code, but since this is already the case for all other `cos-*` scripts I assumed we can go with it for now and once the images and CI stories are addressed raise #108 prio

File: .github/workflows/build-pr.yaml
Comment (mudler) [inline .github/workflows/build-pr.yaml:235]: See here: https://github.com/rancher-sandbox/cOS-toolkit/blob/master/.github/PIPELINES.md, this change needs to be applied in the template, otherwise it would be overwritten each time we regenerate the workflows

File: images/img-builder.sh
Comment (Itxaka) [inline images/img-builder.sh:35]: umm, I missed this but doesnt this means that the raw image will try to boot to `cos` directly and fail? even if we run user data, we need to boot into a "working" system to deploy that data...

**Comment by Itxaka** [comment]
also from my side testing the local raw image seems that now its broken when running it from recovery, stops trying to mount COS_PERSISTENT

File: images/img-builder.sh
Comment (Itxaka) [reply images/img-builder.sh:35]: ah I see, it will keep trying until it falls under cos recovery :+1:

**Comment by Itxaka** [comment]
ahh this seems to be an issue with space, if we dont provide the space needed for the disk expansion to occurr, there doesnt seem to be any errors shown, only the failure to mount cos_persistent...

**Pull request merged**
